crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1915,12 +1927,21 @@ impl<'db> TypeInferenceBuilder<'db> {
 
         if let Some(value) = value.as_deref() {
             let value_ty = self.infer_expression(value);
-            self.add_declaration_with_binding(
-                assignment.into(),
-                definition,
-                annotation_ty,
-                value_ty,
-            );
+            if self.file().is_stub(self.db().upcast()) {
+                self.add_declaration_with_binding(
+                    assignment.into(),
+                    definition,
+                    annotation_ty,
+                    Type::Unknown,
+                );
+            } else {
+                self.add_declaration_with_binding(
+                    assignment.into(),
+                    definition,
+                    annotation_ty,
+                    value_ty,
+                );
+            }
```

Comment:
I don't see any need to duplicate the `add_declaration_with_binding` call here.

And the type we fall back to should be the annotated type, not Unknown. We should add a test showing that after `x: int = ...` in a stub file, we infer `int`, not `Unknown`, for `x`.

```suggestion
            let value_ty = if self.file().is_stub(self.db().upcast()) && value.is_ellipsis_literal_expr() {
                annotation_ty
            } else {
                value_ty
            }
            self.add_declaration_with_binding(
                assignment.into(),
                definition,
                annotation_ty,
                value_ty,
            );
```